### PR TITLE
Metrics overhaul using `torchmetrics`

### PR DESCRIPTION
**Pull request by KnathanM**
This PR seeks to reduce developer burden by switching our loss functions and metrics to `torchmetrics`. This has the following benefits:

* `torchmetrics` is developed by the `lightning` team so they are made to work together, e.g. you can pass an instance of a `torchmetrics.Metric` to `self.log` in a `LightningModule` instead of a raw loss value. 
* `torchmetrics` provides many metrics we don't have. It is straightforward to adapt these to work with our `MPNN` which means we can point people to `torchmetrics` when they want more loss functions/metrics.
* Validation and Test metrics are no longer batch averaged.

This has the following challenges:
* The predictor's criterion are saved as a whole object in model files. This makes them difficult to unpickle if we rename and move class definitions. See below for my work around.
* Classification metrics need to save all predictions and targets in order to calculate metrics like AUROC. This takes more memory than batch averaging the metrics. I don't know if the extra memory used is a problem.

Adapting our loss functions/metrics to `torchmetrics.Metric`s was pretty direct. The `forward` method of `LossFunction` got split into separate `update` and `compute` methods. `Metric` has three key methods. `__init__` sets up internal states that are used to calculate the metric. In MSE for example, these are the squared error and total number of samples. `update` takes the preds and targets and updates the internal states. `compute` uses the internal states to calculate the final metric. The `forward` method is a bit more complicated. It will 
1. save the state of the metric instance
2. reset the metric
3. call `update` and `compute` using the batch to give the loss value for that batch
4. reload the original state of the instance and incorporate the state of the batch's `update` call. 

We use `forward` during training so we can get gradients. During validation we don't need that so we can just call `update`. Then `self.log` will call `compute` to get the metric value to report. 

### Custom unpickle
* The loss functions got moved to `nn.metrics` and some of them got renamed. The unpickler doesn't know where to find the class definition because the location it saved in the model file is no longer there. I found I could intercept what gets passed to the unpickler's `find_class` method by making a custom unpickle class in a custom pickle module. Not sure how stable or advisable that is, but it works. 
* I also had to manually load the checkpoint in `load_from_checkpoint` because I couldn't find how to specify the custom pickle module to `super().load_from_checkpoint`. That's why I load the checkpoint and then save it to a buffer to be loaded again. 
* Even after the unpickler manages to unpickle the criterion saved in old checkpoint files, those criterion are not initialized properly as `torchmetrics` which give an error about the criterion not have the attribute `_defaults`. I get around this by reinitializing the criterion if needed.

### val_loss vs train_loss_epoch
The reported "val_loss" is now calculated the same as the training loss so that the two values reported on the progress bar should be comparable. 
* I added `on_epoch=True` to calculate the training loss over the whole epoch because otherwise the train loss reported is only for a single batch of data and it may be much higher or lower than average. This doesn't affect training, but may confuse users. Calculating the loss at the end of the epoch requires accumulating the preds/targets for all training batches though, which may be a problem for large training sets... I haven't tested large datasets. But also, I'm not sure if setting `on_epoch=False` in `self.log` in `training_step` stops the accumulation anyways. 
* The "val_loss" isn't exactly the same as the "train_loss_epoch" because during validation, batch norm and dropout are set to eval(). 
* For some predictors, the output of `forward` and `train_step` are different, like whether a `sigmoid` is applied in `BinaryClassficationFFN`. Metrics are given the output of `forward` while loss functions are given the output of `train_step`, so `validation_step` has to run the data through the model twice, once through `forward` for metric calculations and once through `train_step` to calculate "val_loss". 
* Some predictors have the same metric class as both the default criterion and default metric, for example `RegressionFFN` has `MSE`. This means the MSE will be calculated twice for the validation set and reported as both "val_loss" and "val/mse". Maybe that could be optimized better.
* Also note that while the validation metric and test metric both use `_evaluate_batch`, the validation metric will be in the scaled space while the test metric will be in the original space, so they aren't directly comparable (only for regression tasks, classification doesn't get scaled). Making them comparable would either require not scaling the validation targets, in which case "val_loss" and "training_loss" would not be comparable, or unscaling the validation metric value before logging, which seems complicated to get right for every metric. 

### Multiclass metrics
* I assume that `CrossEntropyMetric` was the default metric for `MulticlassClassficationFFN` because we didn't really have other multiclass metrics available. [Online](https://stats.stackexchange.com/questions/370861/why-is-cross-entropy-not-a-common-evaluation-metric-for-model-performance) is seems that people don't use CE as a metric. I switched this to use `MulticlassMCCMetric`.
* To make `MulticlassMCCMetric` more accurate, I added another `1 - MCCLoss` so that a value of 1 does indicate a perfect model. 
* Multiclass metrics are difficult to implement in the CLI because in `torchmetrics` the number of classes is required when creating the metric instance, but we dynamically get the number of classes from the shape of `preds`. Either more multiclass metrics need to be manually added to chemprop using our style, or users need to use a jupyter notebook to train on multiclass metrics. I think generally if users want metrics that aren't in Chemprop, including other multiclass metrics, it is reasonable to ask them to use a jupyter notebook. In the docs notebooks for loss functions and metrics, I added an example of adapting multiclass metrics from `torchmetrics` to work with chemprop, to be a resource we can point users to.

### MCC 
This metric is included in `torchmetrics` but that version doesn't allow for task or data weights, so I didn't replace ours. I did have to modify it a bit though to accommodate task weights. `torchmetrics` initializes the metric's states to some default, usually a tensor, in which case, the size of the tensor needs to be known when initialized. We don't require users to say how many tasks they have however. So instead I set the default to an empty list which can hold any size tensor. This is less memory efficient, but means less code changes. 

### metrics as a `ModuleList`
* We have had issues with the metrics not being children of the `MPNN` class because then any of their parameters don't get moved between devices automatically. The simple solution is to make the list of metrics a `ModuleList` instead. 
* One complication of this is then any buffers of the metrics, i.e. the `task_weights` get saved to the state_dict. Old checkpoints are not compatible with this change because Chemprop will now expect there to be task weights for the (arbitrary number of) metrics in the state dict. 
* Currently we register the task weights as a buffer so that they get saved in the state dict. This isn't necessary because we also save the user's task weights (numpy array or `None`) in the hyperparameters in each model.pt file. My preference is to simplify the state_dict by not saving the task weights there, which requires removing them from the state dict for the predictor's criterion when loading old checkpoint. 

### Tracking metric
Previously the first metric in the list of metrics would be used for checkpointing and early stopping. Now this defaults to `val_loss` but can be changed by the user.

### Tests
We are using `torchmetrics` correctly if changing the batch size and shuffling the data doesn't change the metric calculation across the dataset. I test this in `test_metrics.py`. This file has example targets and predictions for every predictor for both `train_step` and `forward`. It then fake trains a model and captures the output of the progress bar to check that "train_loss_epoch" equals "val_loss" and that the validation metric equals the test metric. 
* This also ensures that all of the loss functions/metrics don't have run time errors. 
* It does not test that the metrics are computed correctly. 

`test_loss_functions.py` tests if some metrics are computed correctly.
| Metric | Tested? |
| --- | --- |
| MSE | Implicitly by testing BoundedMSE | 
| MAE | Not tested, seems trivial? |
| RMSE | Not tested but just MSE with a `sqrt()` |
| BoundedMSE | Tested |
| BoundedMAE | The bounded part is tested by BoundedMSE | 
| BoundedRMSE | " " |
| R2Score | Inherited from `torchmetrics` which tests it |
| MVELoss | Tested |
| Evidential | Tested |
| BCELoss | Tested |
| CrossEntropyLoss | Tested |
| BinaryMCCLoss | Tested | 
| BinaryMCCMetric | Implicitly by BinaryMCCLoss |
| MulticlassMCCLoss | Tested | 
| MulticlassMCCMetric | Implicitly by MulticlassMCCLoss |
| BinaryAUROC | Inherited from `torchmetrics` which tests it |
| BinaryAUPRC | Not tested, I have manually checked it before... |
| BinaryAccuracy | Inherited from `torchmetrics` which tests it |
| BinaryF1Score | Inherited from `torchmetrics` which tests it |
| BinaryDirichletLoss | Tested | 
| MulticlassDirichletLoss | Tested |
| SID | Tested |
| Wasserstein | Tested |

I had to change the expected value for SID and Wasserstein for one of the tests because I don't think `threshold` was implemented correctly. The spectra have to only have positive values, which is ensured by using `exp()` on the predictions, but for targets, the user should be able to supply a `threshold` that is a min value for the targets which in v1 defaulted to `1e-8`. See the v1 [README](https://github.com/chemprop/chemprop/blob/02c760216ec03488b69b41abe184589b9c996d74/README.md?plain=1#L333) and this [line](https://github.com/chemprop/chemprop/blob/02c760216ec03488b69b41abe184589b9c996d74/chemprop/train/run_training.py#L188) of code. We still need to make this available to CLI users, but that isn't within the scope of this PR I think. 

### Dirichlet
I think `BinaryDirichletFFN`'s `forward` currently gives logits when it should give probabilities, so I added a sigmoid to that. I didn't understand the v1 code well enough to justify this, but the `forward` of `BinaryClassificationFFN` gives probabilities. `beta / alpha + beta` is not bounded between 0 and 1. I think @hwpang was taking a look at Dirichlet things. Did you find out anything by reading the original paper?

### Other changes
* I removed `validation_loss_function` as it wasn't used and I didn't want to update it.
* I simplified loading the model file in `load_from_checkpoint` and `load_from_file` into one place.

Side note: It can be hard to tell what changes I made to the loss functions because they got moved to a different file. I moved them in commit `48b4659`, so maybe look at the commits up to that point when reviewing that. Most of the loss functions use `_calc_unreduced_loss` so I didn't have to make changes to their code. Only the MCC losses had to see big changes, which I did in commit `d281474`.

### CODE REVIEW

**Comment by KnathanM** [comment]
As part of this we also want to make sure that we add an example of how to use custom loss functions. #772

**Comment by KnathanM** [comment]
The windows test fails for reasons addressed in #1016. I was planning to merge this PR into v2/uncertainty because some metrics are closely tied to uncertainty methods. We should probably merge `main` into `v2/uncertainty` and then `v2/uncertainty` into this branch (`overhaul_metrics`). What do you think @shihchengli?

**Comment by davidegraff** [comment]
this PR touches over **2000** lines of code. It is simply not possible to review such a large PR effectively. Can you think of a way to break this into a stack of PRs?

**Comment by KnathanM** [comment]
I have split this PR into:
#1023
#1025
#1020 -> #1022 -> #1024
